hack/generate_tests.py

```diff
@@ -0,0 +1,157 @@
+          if f == "kustomization.yaml":
+            changed_dirs.add(root)
+
+  return changed_dirs
+
+def find_kustomize_dirs(root):
+  """Find all kustomization directories in root"""
+
+  changed_dirs = set()
+  for top in os.listdir(root):
+    if top.startswith("docs"):
+      print("donotsubmit")
+
+    if top.startswith("."):
+      logging.info("Skipping directory %s", os.path.join(root, top))
+      continue
+
+    if top in TOP_LEVEL_EXCLUDES:
+      continue
+
+    for child, _, files in os.walk(os.path.join(root, top)):
+      for f in files:
+        if f == "kustomization.yaml":
+          changed_dirs.add(child)
+
+  return changed_dirs
+
+def remove_unmatched_tests(repo_root, package_dirs):
+  """Remove any tests that don't map to a kustomization.yaml file.
+
+  This ensures tests don't linger if a pakage is deleted.
+  """
+
+  # Create a set of all the expected test names
+  expected_tests = set()
+
+  for d in package_dirs:
+    expected_tests.add(generate_test_name(repo_root, d))
+
+  tests_dir = os.path.join(repo_root, "tests")
+  for name in os.listdir(tests_dir):
+    if not name.endswith("_test.go"):
+      continue
+
+    # TODO(jlewi): we should rename and remove "_test.go" from the filename.
+    if name in ["kusttestharness_test.go"]:
+      logging.info("Ignoring %s", name)
+      continue
+
+    if not name in expected_tests:
+      test_file = os.path.join(tests_dir, name)
+      logging.info("Deleting test %s; doesn't map to kustomization file",
+                   test_file)
+      os.unlink(test_file)
+
+if __name__ == "__main__":
+  logging.basicConfig(
+      level=logging.INFO,
+      format=('%(levelname)s|%(asctime)s'
+              '|%(pathname)s|%(lineno)d| %(message)s'),
+      datefmt='%Y-%m-%dT%H:%M:%S',
+  )
+  logging.getLogger().setLevel(logging.INFO)
+
+  parser = argparse.ArgumentParser()
+
+  parser.add_argument(
+      "--all",
+      dest = "all_tests",
+      action = "store_true",
+      help="Regenerate all tests")
+
+  parser.set_defaults(all_tests=False)
+
+  args = parser.parse_args()
+
+  repo_root = subprocess.check_output(["git", "rev-parse", "--show-toplevel"])
+  repo_root = repo_root.decode()
+  repo_root = repo_root.strip()
+
+  # Get a list of package directories
+  package_dirs = find_kustomize_dirs(repo_root)
+
+  remove_unmatched_tests(repo_root, package_dirs)
+
+  if args.all_tests:
+    changed_dirs = package_dirs
+  else:
+    changed_dirs = get_changed_dirs()
+
+  for full_dir in changed_dirs:
+    test_name = generate_test_name(repo_root, full_dir)
+    logging.info("Regenerating test %s for %s ", test_name, full_dir)
+
+    test_path = os.path.join(repo_root, "tests", test_name)
+
+    test_target_name = os.path.splitext(test_name)[0]
+    test_target_name = test_target_name.replace("-", "")
+    with open(test_path, "w") as test_file:
+      subprocess.check_call(["./hack/gen-test-target.sh", full_dir,
+                             test_target_name],
```

Comment:
I'm not able to understand the use of this variable 'test_target_name'. Could you please explain?

`gen-test-target` uses only 1 argument as of now, if I'm not wrong.

Reply:
Thanks for catching this. Its a bug; will revert in my next commit.